### ISSUE TITLE
Add MXFP4 MoE/attention backward kernels

### ISSUE THREAD

**Issue by Rocketknight1** [huggingface/transformers#40170]
## A Call To Action!

The Hugging Face community needs you! 🫵 

The `GPT-OSS` model was recently added using `mxfp4` weights. These 4-bit weights are tiny and highly performant on H100/B100/50xx GPUs, but we only have kernels for **forward** passes. This means that users can't train `GPT-OSS` unless they convert the weights to `bfloat16`. This uses 4X more memory, and reduces speed enormously!

We want **native** `mxfp4` training, but that means we need **backward** kernels too. Ideally, these kernels should also support GPUs that don't have FP4 hardware, so users can still benefit from reduced memory usage during training, even if the computation has to be done in FP8 or `bfloat16`.

### Custom Kernels: An extremely short introduction

`transformers` uses the [Kernels](https://github.com/huggingface/kernels) library to load custom kernels. The critical kernel is the MoE kernel, because attention weights are stored in `bfloat16` for `GPT-OSS`.  This kernel lives on the Hub, in the [triton-kernels repo](https://huggingface.co/kernels-community/triton_kernels). The forward kernel is in the [matmul_ogs file](https://huggingface.co/kernels-community/triton_kernels/blob/main/torch-ext/triton_kernels/matmul_ogs.py), but the backward kernel should probably go in its own file.

This won't be easy - the kernels are written in [Triton](https://triton-lang.org/main/index.html), so this is not a good issue for beginners! This will probably be overwhelming if you don't have some experience with raw CUDA or Triton programming already.

Once the hard work of writing the kernel is done, we can help with integrating it into Transformers (there's a [work-in-progress blogpost](https://github.com/huggingface/blog/pull/2971) about this). We can help with serious attempts, but please don't just throw a code agent at the problem or something (you can use a code agent to help with writing and testing the kernel, but only if you're competent enough to evaluate and bugfix its outputs!)

**Comment by tohskai** [comment]
As I said in the twitter post:

I'd be happy to help, but that would require rewriting both forwards and backwards or assume high-level activation checkpointing.

also attention implementation uses old TMA semantics (make block_ptr), i think they will be deprecated eventually.

So the thing is -- i don't think it's very hard to make it work, but fusing and optimizing it well might take some time.

@Rocketknight1

**Comment by akacmazz** [comment]
I've been working on this issue and just opened PR #40180 that addresses the mxfp4 training memory issue.

The 4X memory overhead when converting to bfloat16 for training was really bothering me, so I spent some time implementing the foundational infrastructure needed for native mxfp4 training.

The PR includes:
  - Configuration support for enabling mxfp4 training mode
  - PyTorch autograd integration with custom backward functions
  - Hardware-aware fallback for different GPU generations

While the actual Triton backward kernels still need to be implemented in the `triton-kernels` repository, this provides all the Transformers-side plumbing needed to make it work seamlessly once those kernels are ready.

Would love to get feedback from the maintainers and the community.

Thanks for opening this issue

**Comment by mananchawla2005** [comment]
is it okay to write cuda kernel? or we absolutely need triton?

**Comment by Rocketknight1** [comment]
@mananchawla2005 I think we prefer Triton for now! Do you think there are CUDA optimizations that won't be possible in Triton?

And @tohskai, if you're okay with it, want to try writing the backward kernel for now until I can grab the relevant people to figure out what to do about the forward kernel and the TMA semantics? (I realized after posting this that today is a [public holiday](https://en.wikipedia.org/wiki/Assumption_of_Mary#Public_holidays) in a lot of countries, so we're missing some people right now)

**Comment by ArthurZucker** [comment]
- cuda kernels vs triton: I don't think we have a preference at all @Rocketknight1 now that we are using [`kernels`](https://github.com/huggingface/kernels) in transformers! 
- we can of course rewrite the forward 😉 

As long as the new forward produces the correct outputs, anything is permitted! 🤗

**Comment by tohskai** [comment]
> [@mananchawla2005](https://github.com/mananchawla2005) I think we prefer Triton for now! Do you think there are CUDA optimizations that won't be possible in Triton?
> 
> And [@tohskai](https://github.com/tohskai), if you're okay with it, want to try writing the backward kernel for now until I can grab the relevant people to figure out what to do about the forward kernel and the TMA semantics? (I realized after posting this that today is a [public holiday](https://en.wikipedia.org/wiki/Assumption_of_Mary#Public_holidays) in a lot of countries, so we're missing some people right now)

Sounds reasonable to me. I will start working on this. Is there a way to communicate besides here? Would prefer to ask question if need arises (for example, maybe on numerics, it can be a bit tricky)

**Comment by Rocketknight1** [comment]
@tohskai here should work for now, and I can ping the relevant people at HF if needed!

**Comment by cmp-nct** [comment]
Aside of the problem I ran into, which requires me to remove the assertionto use the kernel on my 5090 (as described in the bug report RK1 has mentioned) I found it not easy to hook into the activations of expert weights because of their fused nature.
It would be nice if it was possible to have a way to access the tensors of the experts.

**Comment by Rocketknight1** [comment]
We can leave things like that for a subsequent PR! The focus here should just be on getting MXFP4 working well in both the forwards and backwards directions

**Comment by Opdoop** [comment]
When will backward be ready? Can to wait to see!

**Comment by avihu111** [comment]
Also super interested in trying it out!

**Comment by Rocketknight1** [comment]
@tohskai have you been able to work on this, or should we try to get other people to take it over?

**Comment by tohskai** [comment]
> [@tohskai](https://github.com/tohskai) have you been able to work on this, or should we try to get other people to take it over?

Sorry, it's a bit too much for me right now, since a lot of corner cases is needed. It would be good if someone else could take over. If you don't find anyone, I'll try again when I am a bit more free, but would hate to be blocker here.

**Comment by Rocketknight1** [comment]
No problem! I'll see if anyone else wants to try it

**Comment by jamesatkindo** [comment]
Super noob question here -- presumably someone at OpenAI had to write these backward kernels to train the gpt-oss models in the first place. Are they not existing within triton already? Also, you mention "native" mxfp4 training -- but this is not possible with H100s I think as those tensor cores do not support fp4, only B200s support native fp4 training on the tensor cores, no? So is this issue about software emulation of mxfp4 training on H100s or native mxfp4 training on B200s? Or maybe I am totally off the mark?
